### PR TITLE
Fix ICE when printing an error message related to mappings

### LINKED ISSUE #12066: [TypeChecker] ICE in UserDefinedValueType::canonicalName()
```
type T0 is bool;
library L {
  function f(mapping(T0=>T0) memory) {}
}
```

throws

https://github.com/ethereum/solidity/blob/0e7e936fe6d49afeec45362cbd228b635833d983/libsolidity/ast/Types.h#L1144

### PR DESCRIPTION
**Pull request by Marenz**
fixes #12066

### CODE REVIEW

File: test/libsolidity/syntaxTests/types/mapping/user_defined_types_mapping_storage.sol [hunk 1]
```diff
@@ -0,0 +1,6 @@
+type A is uint;
+type B is uint;
+library L {
+  function f(mapping(A=>B) storage) public {}
+  function f(mapping(uint=>uint) storage) public {}
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/types/mapping/user_defined_types_mapping_storage.sol]: Can you turn this into a semantic test?

File: libsolidity/ast/Types.h [hunk 2]
```diff
@@ -1141,7 +1141,6 @@ class UserDefinedValueType: public Type
 	}
 
 	std::string toString(bool _short) const override;
-	std::string canonicalName() const override { solAssert(false, ""); }
```
Comment (chriseth) [inline libsolidity/ast/Types.h:1144]: Can you explicitly implement it like it is done for the enum type?

File: Changelog.md [hunk 3]
```diff
@@ -11,6 +11,7 @@ Compiler Features:
 Bugfixes:
  * Commandline Interface: Fix extra newline character being appended to sources passed through standard input, affecting their hashes.
  * SMTChecker: Fix internal error in magic type access (``block``, ``msg``, ``tx``).
+ * TypeChecker: Fix internal error when printing an error message when using a mapping with a user defined value type.
```
Comment (chriseth) [inline Changelog.md]: This is not about the error message.

Comment (Marenz) [reply Changelog.md]: but it's how we trigger it

**Review by chriseth — APPROVED** [review]

**Comment by cameel** [comment]
@Marenz Which issue was this supposed to close? The #11273 you listed in the description is a PR adding fmtlib :)

**Comment by chriseth** [comment]
It was https://github.com/ethereum/solidity/issues/12066

**Comment by Marenz** [comment]
How did that happen Oo thank you for correcting.

**Pull request merged**